File: epi_judge_cpp_solutions/matrix_enclosed_regions.cc
#include <deque>
#include <queue>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>

#include "test_framework/generic_test.h"
#include "test_framework/timed_executor.h"

using std::deque;
using std::pair;
using std::queue;
using std::string;
using std::vector;

void MarkBoundaryRegion(int, int, vector<vector<char>>*);

void FillSurroundedRegions(vector<vector<char>>* board_ptr) {
  vector<vector<char>>& board = *board_ptr;
  // Identifies the regions that are reachable via white path starting from
  // the first or last columns.
  for (int i = 0; i < size(board); ++i) {
    MarkBoundaryRegion(i, /*j=*/0, board_ptr);
    MarkBoundaryRegion(i, size(board[i]) - 1, board_ptr);
  }
  // Identifies the regions that are reachable via white path starting from
  // the first or last rows.
  for (int j = 0; j < size(board.front()); ++j) {
    MarkBoundaryRegion(/*i=*/0, j, board_ptr);
    MarkBoundaryRegion(size(board) - 1, j, board_ptr);
  }

  // Marks the surrounded white regions as black.
  for (vector<char>& row : board) {
    for (char& c : row) {
      c = c != 'T' ? 'B' : 'W';
    }
  }
}

void MarkBoundaryRegion(int i, int j, vector<vector<char>>* board_ptr) {
  queue<pair<int, int>> q(deque<pair<int, int>>(1, {i, j}));
  vector<vector<char>>& board = *board_ptr;
  // Uses BFS to traverse this region.
  while (!empty(q)) {
    const auto [x, y] = q.front();
    q.pop();
    if (x >= 0 && x < size(board) && y >= 0 && y < size(board[x]) &&
        board[x][y] == 'W') {
      board[x][y] = 'T';
      q.emplace(x - 1, y);
      q.emplace(x + 1, y);
      q.emplace(x, y + 1);
      q.emplace(x, y - 1);
    }
  }
}

vector<vector<string>> FillSurroundedRegionsWrapper(
    TimedExecutor& executor, vector<vector<string>> board) {
  vector<vector<char>> char_vector;
  char_vector.resize(board.size());
  for (int i = 0; i < board.size(); i++) {
    for (const string& s : board[i]) {
      if (s.size() != 1) {
        throw std::runtime_error("String size is not 1");
      }
      char_vector[i].push_back(s[0]);
    }
  }

  executor.Run([&] { FillSurroundedRegions(&char_vector); });

  board.clear();
  board.resize(char_vector.size(), {});
  for (int i = 0; i < board.size(); i++) {
    for (char c : char_vector[i]) {
      board[i].emplace_back(1, c);
    }
  }

  return board;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"executor", "board"};
  return GenericTestMain(
      args, "matrix_enclosed_regions.cc", "matrix_enclosed_regions.tsv",
      &FillSurroundedRegionsWrapper, DefaultComparator{}, param_names);
}

File: epi_judge_java_solutions/epi/MatrixEnclosedRegions.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

import java.util.ArrayDeque;
import java.util.List;
import java.util.Queue;

public class MatrixEnclosedRegions {

  public static void fillSurroundedRegions(List<List<Character>> board) {

    // Identifies the regions that are reachable via white path starting from
    // the first or last columns.
    for (int i = 0; i < board.size(); ++i) {
      markBoundaryRegion(i, /*j=*/0, board);
      markBoundaryRegion(i, board.get(i).size() - 1, board);
    }
    // Identifies the regions that are reachable via white path starting from
    // the first or last rows.
    for (int j = 0; j < board.get(0).size(); ++j) {
      markBoundaryRegion(/*i=*/0, j, board);
      markBoundaryRegion(board.size() - 1, j, board);
    }

    // Marks the surrounded white regions as black.
    for (int i = 0; i < board.size(); ++i) {
      for (int j = 0; j < board.get(i).size(); ++j) {
        board.get(i).set(j, board.get(i).get(j) != 'T' ? 'B' : 'W');
      }
    }
  }

  private static class Coordinate {
    public Integer x;
    public Integer y;

    public Coordinate(Integer x, Integer y) {
      this.x = x;
      this.y = y;
    }
  }

  private static void markBoundaryRegion(int i, int j,
                                         List<List<Character>> board) {
    Queue<Coordinate> q = new ArrayDeque<>();
    q.add(new Coordinate(i, j));
    // Uses BFS to traverse this region.
    while (!q.isEmpty()) {
      Coordinate curr = q.poll();
      if (curr.x >= 0 && curr.x < board.size() && curr.y >= 0 &&
          curr.y < board.get(curr.x).size() &&
          board.get(curr.x).get(curr.y) == 'W') {
        board.get(curr.x).set(curr.y, 'T');
        q.add(new Coordinate(curr.x - 1, curr.y));
        q.add(new Coordinate(curr.x + 1, curr.y));
        q.add(new Coordinate(curr.x, curr.y - 1));
        q.add(new Coordinate(curr.x, curr.y + 1));
      }
    }
  }

  @EpiTest(testDataFile = "matrix_enclosed_regions.tsv")
  public static List<List<Character>>
  fillSurroundedRegionsWrapper(List<List<Character>> board) {
    fillSurroundedRegions(board);
    return board;
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "MatrixEnclosedRegions.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}
